src/sentry/grouping/enhancer.py

```diff
@@ -0,0 +1,299 @@
+        self.key = key
+        self.pattern = pattern
+
+    def _to_config_structure(self):
+        return MATCH_KEYS[self.key] + self.pattern
+
+    @classmethod
+    def _from_config_structure(cls, obj):
+        return cls(SHORT_MATCH_KEYS[obj[0]], obj[1:])
+
+
+@implements_to_string
+class Action(object):
+
+    def __init__(self, key, flag, range):
+        self.key = key
+        self.flag = flag
+        self.range = range
+
+    def __str__(self):
+        return '%s%s%s' % (
+            {'up': '^', 'down': 'v'}.get(self.range, ''),
+            self.flag and '+' or '-',
+            self.key,
+        )
+
+    def _to_config_structure(self):
+        return ACTIONS.index(self.key) | (ACTION_FLAGS[self.flag, self.range] << 5)
+
+    @classmethod
+    def _from_config_structure(cls, num):
+        flag, range = REVERSE_ACTION_FLAGS[num >> 5]
+        return cls(ACTIONS[num & 0xf], flag, range)
+
+
+class Enhancements(object):
+
+    def __init__(self, rules, changelog=None, version=None, bases=None, id=None):
+        self.id = id
+        self.rules = rules
+        self.changelog = changelog
+        if version is None:
+            version = VERSION
+        self.version = version
+        if bases is None:
+            bases = []
+        self.bases = bases
+
+    def as_dict(self, with_rules=False):
+        rv = {
+            'id': self.id,
+            'changelog': self.changelog,
+            'bases': self.bases,
+        }
+        if with_rules:
+            rv['rules'] = [x.as_dict() for x in self.rules]
+        return rv
+
+    def _to_config_structure(self):
+        return [self.version, self.bases, [x._to_config_structure() for x in self.rules]]
+
+    def dumps(self):
+        return base64.urlsafe_b64encode(msgpack.dumps(
+            self._to_config_structure()).encode('zlib')).strip('=')
+
+    def iter_rules(self):
+        for base in self.bases:
+            base = ENHANCEMENT_BASES.get(base)
+            if base:
+                for rule in base.iter_rules():
+                    yield rule
+        for rule in self.rules:
+            yield rule
+
+    @classmethod
+    def _from_config_structure(cls, data):
+        version, bases, rules = data
+        return cls([Rule._from_config_structure(x) for x in rules], version, bases)
+
+    @classmethod
+    def loads(cls, data):
+        padded = data + b'=' * (4 - (len(data) % 4))
+        try:
+            return cls._from_config_structure(msgpack.loads(
+                base64.urlsafe_b64decode(padded).decode('zlib')))
+        except (LookupError, AttributeError, TypeError, ValueError) as e:
+            raise ValueError('invalid grouping enhancement config: %s' % e)
+
+    @classmethod
+    def from_config_string(self, s, bases=None, id=None):
+        try:
+            tree = enhancements_grammar.parse(s)
+        except ParseError as e:
+            context = e.text[e.pos:e.pos + 33]
+            if len(context) == 33:
```

Comment:
What is special about 33?

Reply:
Nothing. I wanted to trim it to 32 characters and 33 is the one extra char to determine if I need to add an ellipsis marker.